File: src/Worker/Worker.cs
using Microsoft.Extensions.Options;
using Worker.Models;
using Worker.Services;
using Worker.Settings;

namespace Worker;

public class Worker : BackgroundService
{
    private readonly ILogger<Worker> _logger;
    private readonly GlobalSettings _globalSettings;
    private readonly FFMpegService _ffMpegService;
    private readonly InputStorageService _inputStorageService;
    private readonly OutputStorageService _outputStorageService;

    public Worker(ILogger<Worker> logger, IOptions<GlobalSettings> globalSettings, FFMpegService ffMpegService,
        InputStorageService inputStorageService, OutputStorageService outputStorageService)
    {
        _logger = logger;
        _ffMpegService = ffMpegService;
        _inputStorageService = inputStorageService;
        _outputStorageService = outputStorageService;
        _globalSettings = globalSettings.Value;
    }

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        while (!stoppingToken.IsCancellationRequested)
        {
            try
            {
                await RunAsync(stoppingToken);
            }
            catch (Exception e)
            {
                _logger.LogError(e, "Error during Worker execution:");
            }

            await Task.Delay(_globalSettings.Delay, stoppingToken);
        }
    }

    private async Task RunAsync(CancellationToken cancellationToken)
    {
        var message = await _inputStorageService.ReceiveMessageAsync<InputMessage>(cancellationToken);

        if (message == null)
        {
            return;
        }

        var inputFilePath = await _inputStorageService.DownloadBlobAsync(message.BlobName, cancellationToken);

        var conversionResult = await _ffMpegService.ConvertAsync(inputFilePath, message.Arguments, message.TargetExtension);

        if (conversionResult != null)
        {
            await _outputStorageService.UploadBlobAsync(conversionResult.Name, conversionResult.FullName, cancellationToken);

            var resultQueueMessage = new OutputMessage(message.Id, conversionResult.Name);

            await _outputStorageService.SendMessageAsync(resultQueueMessage, cancellationToken);
        }
    }
}

File: src/Worker/Worker.fs
namespace Worker

open System.Diagnostics
open System.Threading
open System.Threading.Tasks
open Domain
open Domain.Repos
open Domain.Settings
open Infra
open Infra.Helpers
open Infra.Queue
open Microsoft.Extensions.Hosting
open Microsoft.Extensions.Logging
open Microsoft.Extensions.Options
open OpenTelemetry
open OpenTelemetry.Context.Propagation
open Worker.Settings
open Domain.Workflows
open FsToolkit.ErrorHandling

type Worker
  (
    logger: ILogger<Worker>,
    workerSettings: WorkerSettings,
    inputQueue: IInputQueue,
    getOutputQueue: GetOutputQueue,
    inputStorage: IInputStorage,
    outputStorage: IOutputStorage,
    converter: IConverter,
    appOptions: IOptions<AppSettings>
  ) =
  inherit BackgroundService()

  let processQueueMessage (queueMessage: QueueMessage) =
    let inputMessage =
      JSON.deserialize<BaseMessage<{| Id: string; Name: string |}>> queueMessage.Body

    let request: Conversion.Request =
      { Id = inputMessage.Data.Id
        Name = inputMessage.Data.Name }

    let inputMsgClient =
      inputQueue.GetInputMsgClient(queueMessage.Id, queueMessage.PopReceipt)

    let outputQueue = getOutputQueue (inputMessage.Context, inputMessage.Data.Id)

    let convert =
      Conversion.run inputStorage outputStorage inputMsgClient outputQueue appOptions converter LocalStorage.deleteFile

    task {
      let parentContext =
        Propagators.DefaultTextMapPropagator.Extract(
          PropagationContext(),
          inputMessage.Context,
          fun d key ->
            match d.TryGetValue key with
            | true, value -> [ value ]
            | _ -> []
        )

      Baggage.Current <- parentContext.Baggage

      use activity =
        Observability.ActivitySource.StartActivity("Convert", ActivityKind.Consumer, parentContext.ActivityContext)

      try
        do! convert request
      with e ->
        logger.LogError(e, "Error during processing queue message")
        do! inputMsgClient.Delete()
        do! outputQueue.SendFailureMessage()
    }

  let runWorker () =
    inputQueue.GetMessage()
    |> Task.bind (function
      | Some m -> processQueueMessage m
      | None -> Task.FromResult())

  override _.ExecuteAsync(ct: CancellationToken) = task {
    while not ct.IsCancellationRequested do
      try
        do! runWorker ()
      with e ->
        logger.LogError(e, "Worker error")

      do! Task.Delay(workerSettings.Delay)
  }